Add forward AD formulas for mv, scatter_add, _s_where

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71026
* __->__ #70468
* #70355

Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

soulitzer committed:
Add forward AD formulas for mv, scatter_add, _s_where

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"




[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1634,6 +1636,7 @@
   condition: non_differentiable
   self: where(condition, grad, zeros_like(grad))
   other: where(condition, zeros_like(grad), grad)
+  result: where(condition, self_t, zeros_like(other_p)) + where(condition, zeros_like(self_p), other_t)
```
Why not `where(condition, self_t, other_t)`?

soulitzer replied:
!

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"




[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405364).

albanD reviewed (approved):
SGTM

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -1072,6 +1072,7 @@
 - name: mv(Tensor self, Tensor vec) -> Tensor
   self: grad.ger(vec.conj())
   vec: self.conj().t().mv(grad)
+  result: mv(self_t, vec_p) + mv(self_p, vec_t)
```
it's not necessary, but I think it would be cool to add a ZeroTensor fastpath for `mv` as a follow-up.

albanD replied:
Yes!

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405364).

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405364).

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405364).

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405364).

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405364).

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for mv, scatter_add, _s_where"


Differential Revision: [D33405364](https://our.internmc.facebook.com/intern/diff/D33405364)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405364).

Lezcano commented:
Now that I see this, I had a PR that's completely stalled due to a mobile error that implemented a number of forward formulas, finally enabling forward AD for matmul and a few other functions. See https://github.com/pytorch/pytorch/pull/64387
